kernel/src/expressions/column_name.rs

```diff
@@ -0,0 +1,313 @@
+use std::borrow::Borrow;
+use std::fmt::{Display, Formatter};
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+/// A (possibly nested) column name.
+// TODO: Track name as a path rather than a single string
+#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord)]
+pub struct ColumnName {
+    path: String,
+}
+
+impl ColumnName {
+    /// Constructs a new column name from an iterator of field names. The field names are joined
+    /// together to make a single path.
+    pub fn new(path: impl IntoIterator<Item = impl Into<String>>) -> Self {
+        let path: Vec<_> = path.into_iter().map(Into::into).collect();
+        let path = path.join(".");
+        Self { path }
+    }
+
+    /// Joins two column names, concatenating their fields into a single nested column path.
+    pub fn join(left: ColumnName, right: ColumnName) -> ColumnName {
+        [left, right].into_iter().collect()
+    }
```

Comment:
I think either this should take `&self` and then work as if self was `left` and join to the arg (similar to `Url`'s [join](https://docs.rs/url/latest/url/struct.Url.html#method.join)), or it should live outside the `impl`.

I'd prefer:
```suggestion
    pub fn join(&self, right: ColumnName) -> ColumnName {
        [self.clone(), right].into_iter().collect()
    }
```

Reply:
How could it live outside the impl without causing massive namespace collisions? `join` is a pretty generic name and rust doesn't have function overloading...

I was avoiding the `&self` join route because it potentially means copying a bunch of strings for deeply nested paths. But (a) that's probably optimizing rather than merely not-pessimizing; and (b) the left side of the join is usually a single field name and we're anyway consuming the right arg.

Reply:
We could also define a `join_with` method that takes `self`. Maybe I'll just do that.

Reply:
Note: [std::path::Path::join](https://doc.rust-lang.org/std/path/struct.Path.html#method.join) actually takes both left and right sides by reference. Let's go simpler-is-better here.